pkg/scheduler/scheduler.go

```diff
@@ -144,19 +185,24 @@ func (s *Scheduler) schedule(ctx context.Context) {
 		cq := snapshot.ClusterQueues[e.ClusterQueue]
 		if cq.Cohort != nil {
 			// Having more than one workloads from the same cohort admitted in the same scheduling cycle can lead
-			// to over admission if:
+			// to over admission if they have any common resources assigned and:
 			//   1. One of the workloads is borrowing, since during the nomination the usage of the other workloads
 			//      evaluated in the same cycle is not taken into account.
 			//   2. An already admitted workload from a different cluster queue is borrowing, since all workloads
 			//      evaluated in the current cycle will compete for the resources that are not borrowed.
-			if usedCohorts.Has(cq.Cohort.Name) && (e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues()) {
-				e.status = skipped
-				e.inadmissibleMsg = "other workloads in the cohort were prioritized"
-				continue
+			if cycleCohortsUsage.hasCommonFlavorResources(cq.Cohort.Name, e.assignment.Usage) && (e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues()) {
```

Comment:
Actually, IIUC, the only condition checked here we really need is
```golang
if  !cq.Cohort.CanFit(cycleCohortsUsage.totalUsageForCommonFlavorResources(cq.Cohort.Name, e.assignment.Usage)) {
```

The other conditions are just performance optimizations. 
1. `cycleCohortsUsage.hasCommonFlavorResources(cq.Cohort.Name, e.assignment.Usage)`

I think we can safely skip. If there is no intersection, then the computation of `totalUsageForCommonFlavorResources` should be fast. I see no reason to complicate the implementation. 

2. `(e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues())` up to you if you want to keep it

It took me a while to figure out which conditions are just performance optimizations. IMO the comment does not currently help cause it pays attention to borrowing as the main condition, while this is just a perf optimization. 

I would suggest to make the comment along the lines:
```
Check if the workload assignment still fits in the cohort.
Additionally, as a performance optimization, skip the check if there is no
borrowing in the cohort. We can do this, because the check was already
performed per cluster queue during the nomination phase.
```

Reply:
Borrowing is the main point of the checks.

 `(e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues())` are both zero allocation methods.

If there is no borrowing going on, there is no point in checking `CanFit` since all workloads will only use resources in available in their cluster queue nominal quota.

Reply:
IIUC the code gives the same results with and without the check, right? 

The only reason for the check is the perf optimization, because it guards extra allocations in some scenarios. As mentioned, I'm ok to keep the check, but the comment should make it clear that this is for perf reasons. Also, leave the discussion open in case @alculquicondor has some view.

Reply:
The checks are documented above, and we can brake down the comment as:
```go
			// Having more than one workloads from the same cohort admitted in the same scheduling cycle can lead
			// to over admission if they have any common resources assigned and:
			//   1. One of the workloads is borrowing, since during the nomination the usage of the other workloads
			//      evaluated in the same cycle is not taken into account.
			//   2. An already admitted workload from a different cluster queue is borrowing, since all workloads
			//      evaluated in the current cycle will compete for the resources that are not borrowed.
```
we get something like:

`if they have any common resources assigned` => `cycleCohortsUsage.hasCommonFlavorResources(cq.Cohort.Name, e.assignment.Usage)`

`1.  One of the workloads ......` => `e.assignment.Borrows() `
`2.  An already admitted workload .....` => `cq.Cohort.HasBorrowingQueues()`

If it is not the case we should not do any other checks.

Reply:
Sure, I just wanted to emphasize, that with the new check, the previous checks are just perf. optimization. I was thinking about something like this:

```golang
			// Check if the workload assignment still fits in the cohort.
			// Additionally, as a performance optimization, skip the check if there is no
			// borrowing in the cohort. We can do this, because the check was already
			// performed per cluster queue during the nomination phase.
			if (e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues()) && !cq.Cohort.CanFit(cycleCohortsUsage.totalUsageForCommonFlavorResources(cq.Cohort.Name, e.assignment.Usage)) {
				e.status = skipped
				e.inadmissibleMsg = "other workloads in the cohort were prioritized"
				continue
			}
```
Let me know what you think. @alculquicondor ?

Reply:
I have another proposal that I think saves on more allocations:

```golang
cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage)
if !cq.Cohort.CanFit(cycleCohorts) {
  e.status = skipped
  e.inadmissibleMsg = "other workloads in the cohort were prioritized"
  continue
}
```

Checking `e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues()` would be accurate, but I don't think it's necessary anymore (unit tests should prove us right or wrong).

Reply:
I like the new proposal, just thinking if we do `cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage)` unconditionally we may say there is no more space, and no more workloads are admitted. In the previous solution, we still would consider the next workload, right?

Reply:
I think the code in the main branch is always more restrictive.

Reply:
IIUC yes. Say, we have three workloads: 
1. accepted with borrowing and consumes some quota
2. rejected because it would exceed the remaining quota, 
3. rejected because the second consumed the "virtual quota" 

Indeed still less strict than the main branch, because in the main branch 3. would be rejected due to the fact of borrowing. However, I think the reasoning with the "virtual quota" required is non trivial.

TBH I think the simplest conceptually is 

```golang
if cq.Cohort.CanFit(cycleCohorts) {
  e.status = skipped
  e.inadmissibleMsg = "other workloads in the cohort were prioritized"
  continue
}
cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage)
```
I'm not sure we would have a performance issue here. Yes, there are some allocations, but we do allocations during each admission cycle anyway, we also take read locks.

Reply:
did you take into consideration that we sort entries that don't require borrowing last?
2 and 3 would need to require borrowing if they were checked after 1.

Reply:
The entries with borrowing go the end: https://github.com/kubernetes-sigs/kueue/blob/624ceee834d65602c7a5a107a91e4b15ace6f839/pkg/scheduler/scheduler_test.go#L1103.

Yes, I took this into account and this is required for your proposal to be less restrictive that "main". This is why I find it a little bit hard to digest.

Still the proposal is more restrictive (unnecessarily) than the proposal from https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1290352459 (or in the current branch), because in the scratched scenario it may reject workload (3.) unnecessarily.

Reply:
In main, both 2 and 3 would be rejected, so my proposal is not more restrictive than that.

But you missed something. It should be (in pseudocode):

`if !cq.Cohort.CanFit(cycleCohortsUsage+currentWorkloadUsage)`

What I propose does potentially less calculations and allocations, while being permissive and restrictive enough. But I'm ok with either solution.

Reply:
Yes, sure I meant `if !cq.Cohort.CanFit(cycleCohortsUsage+currentWorkloadUsage)` :). 

I'm leaning towards it, as being more easy to understand. In particular, your proposal requires the ordering that groups borrowing workloads at the end. This might be a little bit fragile assumption, and easy to miss as we go, IMO.

So basically, the @trasc proposal, but without the extra checks for perf optimizations. I think saving on allocations in this check isn't critical, as on every scheduling cycle we do anyway: 
* We take read locks to 
  - pick top elements for CQs. 
  - create the snapshot
* we anyway do allocations to
  - represent the head workloads
  - represent the nomination information
  - represent and grow the `cycleCohortsUsage` map
* we sort all the entries (nlogn)

Thus, I think the extra allocations to build the `cycleCohortsUsage+currentWorkloadUsage` map will not matter that much. We could also probably try to reuse it instead of doing `cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage)` from scratch.

Reply:
IMO unfortunately the perf optimizations are not without a cost, the cost is hidden in complexity, which can be reflect by the comment above. I think in the proposed solution then comment can be as brief as ` Check if the workload assignment still fits in the cohort.`

Reply:
> We could also probably try to reuse it instead of doing cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage) from scratch. 

I think implementing some kind of 2PC for this will be more complex that the extra if-s. 

Lt's keep this as it is.

Reply:
2PC is a distributed algorithm, so not sure how it is relevant here. 

I meant the following reformulation:
```golang
	nextCycleCohotsUsage := cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage)
	if !cq.Cohort.CanFit(nextCycleCohotsUsage) {
		e.status = skipped
		e.inadmissibleMsg = "other workloads in the cohort were prioritized"
		continue
	}
	cycleCohortsUsage = nextCycleCohotsUsage
```
Note that it avoids the extra allocations related to building the same map twice as in the current branch - once to check `CanFit` (represnted by `sum) and once to do `Add`.  In the current branch, the allocations are saved only in a happy path. In the not so happy path, there is the same amount of allocations, or even more than in the proposal in this comment. 

Still, I think it is a secondary question which proposal we choose out of those listed in comments https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1290282730, https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1290352459, and this one. What they have in common is to avoid the extra checks. As discussed in the comment https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1291003945 we already have allocations and read locks taken in each cycle. 

The cost of the extra checks is code complexity, which means longer dev time in the future, and risk of issues. Finally, it is not even clear which solution is faster. I think I make the argument clear, but I see there might be a trade off, so let me defer the decision to @alculquicondor .

Reply:
> 2PC is a distributed algorithm, so not sure how it is relevant here

I meant something like:
Phase 1: compute the sum (intersection) 
Phase 2: merge the sum in `cycleCohortsUsage`  if the workload goes to admission.

Having `cycleCohortsUsage` fully cloned did not cross my mind.

Reply:
Overall all the proposals are quite similar. To recap:
- Checking `cycleCohortsUsage.hasCommonFlavorResources` that might save allocations only if no flavors are shared.
- Checking "is borrowing" also seems to be a performance optimization given the stronger checks we have now. But we should know from the unit tests.

We don't have benchmarks to inform our decision, but since all the options have pros and cons from a performance perspective, I would prefer to save on code complexity.

I would go for the last proposal https://github.com/kubernetes-sigs/kueue/pull/1039#discussion_r1291111059

It could also be more like a backtrack, to save on allocations in most scenarios.

```golang
cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage)
if !fit {
  cycleCohortsUsage.remove(cq.Cohort.Name, e.assignment.Usage)
  ...
  continue
}
```

Reply:
Sure the are multiple ways of doing it, but as long as the one already done is not wrong , why change?

Reply:
Because of the readability, which affects maintainability

Reply:
Having the something like:
```
cycleCohortsUsage.add(cq.Cohort.Name, e.assignment.Usage)
if !fit {
  cycleCohortsUsage.remove(cq.Cohort.Name, e.assignment.Usage)
  ...
  continue
}
```
Is not sufficient, we still need to accept the first workloads "touching" a resource even if it results in going over the limits, when the workloads requires preemption. Case in which any other workloads after this, using the same resources or not will not fit if we don't filter the resources we check in CanFit.

However I tried to simplify the checks.

Reply:
Right, it would not allow preemption. Can you include this learning in the comment?

I'm almost tempted to ask again about the initial proposal: Only check which flavors are in use. But you added good coverage, so hopefully this should minimize maintenance in the future.

Reply:
done

Reply:
Thanks for applying the simplification and revising the comment. Probably the `hasCommonFlavorResources` check is still extra, but I believe the code is already more readable.